isis/src/control/objs/ControlPoint/ControlPoint.cpp

```
@@ -962,12 +962,14 @@ namespace Isis {
 
     // if point is Free, we continue to compute a priori coordinates
 
-    // if no good measures, we're done
-    // TODO: is the message true/meaningful?
+    // If no good measures, this point is ignored
     if (goodMeasures == 0) {
-      QString msg = "in method ControlPoint::ComputeApriori(). ControlPoint [" + GetId() + "] has ";
-      msg += "no measures which project to the body";
-      throw IException(IException::User, msg, _FILEINFO_);
+      QString msg = "In method ControlPoint::ComputeApriori(): ControlPoint [" + GetId() + "] has ";
+      msg += "no measures which project to the body. Will ignore it.";
+      SetIgnored(true);
+      std::cerr << msg << std::endl;
```

I like the warning but for most users this will still dump to the terminal, or most won't pipe the error message to another output. We need a way to silence the warnings in the preference file or some other setting. @Kelvinrr Thoughts on this? I know we had issues when we added a deprecation warning for users, not sure what discussions came out of that

---

Sometimes many control points will fail to work. I am not sure what you have in mind, but we probably would not want a popup for each of them. This could even be silent, with failed points simply glossed over. Fine with me whichever way this would be dealt with.

---

yeah we almost universally get complaints about warnings, it can mess up people who parse app outputs which is very common since ISIS often returns PVL.

---

Can you suggest what this line should be changed to? There appear to be various choices in several places.

---

Hi All - could someone please be more explicit about this what is being discussed here as this is a program my job depends on and changes have impact. Is this a situation where measures for a point fail to intersect the body? If so, these should not be silenced and the program either needs to fail outright (as it does now) and/or that information should go to a log file. I want to know about these situations as they indicate some sort of a problem and should be removed and/or my network building needs additional checks. More importantly I need to understand why it is happening and hiding the error does not give me that opportunity.

---

With control networks I make in the ASP bundle_adjust program, I believe jigsaw starts to fail when the angle between rays is too narrow. I am not fully sure, but I know that if I filter out ahead of time points based on convergence angle, it goes through.

I agree the user should be told about such issues. I'd rather it not fail. The question is, how to log these properly, without pop-ups for example. 

Can one point out to a few lines of code used for warnings in such situations? Above, code like "PvlGroup radiusSolveWarning("RadiusSolveWarning")" is used, maybe something similar?

---

I don't know anything about specific code to point out possible warning scenarios, but I don't find pop-ups (without failure) useful in the case of invalid control points. Some of my bundles run hours and in one case days, so I would prefer the program fail outright so that I can do something about the offending measures/points as soon as possible. Plus these longer runs are typically sent to a cluster where I am not monitoring messages to screen in real time. Finding out long after the fact that something is broken is not an efficient use of my time. 

I'm not sure how or where this information could be captured in any existing jigsaw report file. Plus I think changes like that or going to warning versus the current program failure is a breaking change (I'm not certain about the latter though). 

I would discourage depending on jigsaw to validate networks in the way that is being suggested since there are programs that already do this that should be run before passing a network to jigsaw. I can see how a measure may get bundled off the target during jigsaw, but I think that is a different scenario.

---

Why programs validate the control network before jigsaw?

And yes, what you point out is correct. Jigsaw itself could create problematic control points that it later itself would not be able to handle. One could have a network with millions of control points and some of the cameras may have borderline acceptable pointing info (such looking in an almost parallel way).

I think it is a sensible choice that if the tool cannot handle a control point, it should be skipped with a warning. Otherwise the user is stuck with a failure, and there seems to be no way around it. The user is expected to inspect warnings and make decisions.

---

> Why programs validate the control network before jigsaw?
cnetstats, cnetcheck

---

Will cnetcheck go through all the same motions as jigsaw? Can then the failure message refer to these two?

---

cnetcheck will report if the measure does not intersect body when nolatlon=True (default) and send the information to an output file (which is not the most user friendly format so you might need a script to parse the output to send to cnetedit to remove the offending measures). It doesn't iterate like jigsaw, it will only check the state of the network based on spice and the line/sample information of the measures (probably similar to what jigsaw does before it starts to iterate). But it will report on off target measures along with free points having single measures, points having nocontrol (because all measures are invalid (ignored), etc. 

cnetcheck is more robust that cnetstats.

It might be useful if jigsaw referred the user to cnetcheck in the case where a measure does not intersect the target body. But I'm not sure the particular bundle error I'm referring to for off-body targets is the same as what you are experiencing with your networks.

I won't go into all of the detail, but this is the jigsaw failure output for voyager data where measures essentially had no spice because they were off the target:

```
  Group = Error
    Program = jigsaw
    Class   = "USER ERROR"
    Code    = 2
    Message = "Unable to bundle adjust network [tb_vgr_voygalnet_Epos.net]"
    File    = main.cpp
    Line    = 157
  End_Group

  Group = Error
    Program = jigsaw
    Class   = "USER ERROR"
    Code    = 2
    Message = "Could not solve bundle adjust"
    File    = BundleAdjust.cpp
    Line    = 982
  End_Group

  Group = Error
    Program = jigsaw
    Class   = "USER ERROR"
    Code    = 2
    Message = "Unable to map apriori surface point for measure
               Voyager2/NAC/2064922 on point 133 into focal plane"
    File    = BundleAdjust.cpp
    Line    = 1887
  End_Group

```
I can't tell if this is the same sort of error you were encountering with your Juno data though.
I will add that for my Voyager and some Galileo disk data of Europa we attempted to avoid this sort of problem by manually moving the spice onto the target for images having issues. Very user intensive, but the process works (I can share those details in an email if you are interested and if applicable to your particular situation).

I have also encountered problems with Galileo SSI Europa, Cassini ISS Enceladus, Titan and MEX SRC and HRSC Phobos data where points end up right on the limb for images (having spice mostly on the target) because a measure is a fraction away of having spice or not. Cnetcheck has found those instances (reported in nolatlon output) and they are removed from the network prior to running the network through the bundle adjustment.  If this is more in line with what you are experiencing with you JunoCam data you might see if cnetcheck spots your known problems.

And so I understand better, in the suggested warning message
```
QString msg = "In method ControlPoint::ComputeApriori(): ControlPoint [" + GetId() + "] has ";
      msg += "no measures which project to the body. Will ignore it.";
      SetIgnored(true);
      std::cerr << msg << std::endl;
```
Does "ignore" mean literally setting a measure in the network to ignore or does it simply mean not using it in calculations during any or all iterations? 

If it is that latter (either really, the user needs to know if jigsaw is modifying the state of measures), maybe it could be flagged as "rejected" which would show up in the residuals.csv file in a similar way as measures are via the use of OUTLIER_REJECTION. But the user would need to know to look in the residuals file for rejected measures and be able to discern rejection due to outlier_rejection use versus rejection due to failing to intersect the target. That seems messy, so maybe ignore that idea. Which is why I think depending on jigsaw to validate a network should be discouraged and cnetcheck should be run to catch some of the most egregious issues before hand.

---

>Does "ignore" mean literally setting a measure in the network to ignore or does it simply mean not using it in calculations during any or all iterations?

The control point will be flagged as ignored in the control network, and will be skipped when forming the optimization problem.

>I will add that for my Voyager and some Galileo disk data of Europa we attempted to avoid this sort of problem by manually moving the spice onto the target for images having issues. Very user intensive, but the process works

This is likely not a great solution. Jigsaw should have the ability to filter out various problematic points without failing. It should print a report of the fraction of points that cannot be processed.

I rarely use this, so will defer to actual users and maintainers, but I think this is a case of a tool that should be more robust and give the user more choices of how to handle borderline data.

---

> This is likely not a great solution. Jigsaw should have the ability to filter out various problematic points without failing. It should print a report of the fraction of points that cannot be processed.

No, not fun, but in the case of Voyager data of Europa there was no choice - the spice was completely off the target and had to be moved. Extreme case. 

I don't completely disagree that jigsaw should be capable of ignoring problem measures, reporting on it and continuing to iterate if possible. However it may be tough parsing the various possible errors (with some level of specificity) into a single new output report.

---

> However it may be tough parsing the various possible errors (with some level of specificity) into a single new output report.

One has to start somewhere. Ignoring failed control points and adding a report file is a good place. But since regular users seem not to see it as a problem, I guess maybe it is fine, most of the time.

BTW, for JunoCam I ran into systematic problems in produced terrain models that were due to intrinsics. ASP can tune those, and we have a [doc](https://stereopipeline.readthedocs.io/en/latest/examples/junocam.html).